surveys/promise-rejections/survey.md

```diff
@@ -0,0 +1,235 @@
+There are two ways to handle rejections: by attaching a `.catch` handler to it,
+or by awaiting on the promise within a try/catch block. In both cases, the
+handling of the rejection (the execution of the callback passed to `.catch`, or
+the execution of the `catch {}` block) will happen in a future turn of the
+event loop.
+
+Promises are designed so that attaching handlers or awaiting can be done at any
+point in time, from when the Promise was created (possibly while it's still
+pending), to right before the program finishes execution.
+
+A rejection is considered unhandled from the point it happens until the point
+where a handler is attached to the Promise or the Promise is awaited within a
+`catch {}` block. Below are a few examples of handled and unhandled rejections.
+
+```js
+async function foo() {
+  throw new Error();
+}
+
+foo()  // 1. Unhandled at this point
+  .catch(() => console.error("an error occured"));  // 2. Now it's handled
+
+try {
+  await foo();
+} catch(e) {  // 3. Handled
+  console.error("an error occured")
+}
+
+foo(); // 4. Unhandled, but execution continues
+
+const rejected = foo(); // 5. Unhandled on current event loop turn
+
+// 4. ... but handled in a future turn of the event loop when the setTimeout
+// callback is executed.
+setTimeout(() => rejected.catch(() => console.error("an error occured")), 100);
+```
+
+As we can see in the examples, an unhandled rejection might be handled in the future, like example 4, but it might also stay unhandled forever (like example 5).
+
+Certain unhandled rejections may in rare cases leave your application in a non-deterministic and unsafe state, whether it's internal application state (including memory leaks), external resources used by your application (say, file handles or database connections), or external state (say, consistency of data in a database).
+
+As an example, the following server is not sending a response back to the client, causing a socket leak and a possible Denial of Service attack:
+
+```js
+const http = require('http')
+const server = http.createServer(handle)
+
+server.listen(3000)
+
+function handle (req, res) {
+  doStuff()
+    .then((body) => {
+      res.end(body)
+    })
+}
+
+function doStuff () {
+  if (Math.random() < 0.5) {
+    return Promise.reject(new Error('kaboom'))
+  }
+
+  return Promise.resolve('hello world')
+}
+```
+
+## Are you currently using Promises, async functions, a mix, or neither?
+
+(Multiple choice)
+
+  - [ ] Native Promises (`new Promise`, `Promise.resolve`, `Promise.reject`)
+  - [ ] Third-party Promises (for example: Bluebird)
+  - [ ] async functions
+  - [ ] Not writing Promise-based code or using Promise-based libraries
+
+## How are you handling rejections today?
+
+When consuming Promises, async functions or thenables, which of the options below do you use to handle rejections?
+
+(Multiple choice)
+
+  - [ ] `.catch()`
+  - [ ] `try / catch` wrapping an `await` operation
+  - [ ] Leave the handling to my caller
```

Comment:
Do we want to consider an extra option

```
 - [ ] Add `process.on('unhandledRejection', handleError)` and handle errors in a single place.
```

This is different then `Leave the handling to my caller` because you can then write code where you do

```js
async function doSomething() {}`

doSomething() // no await / catch;
// or more realistically `x.on('someevent', doSomething)`
```

The caller does not have `.catch()` nor `try / catch` ;

This is mostly about using `async / await` without `try / catch / .catch()` using a library like https://github.com/scopsy/await-to-js

Reply:
I still think this is covered in the "leave the handling to the caller" (maybe changing to "leave the handling to someone else"?) and (especially) by the other question. Adding an option for unhandledRejection handler here would mix localized handling with global handling, and IMO that's not the goal of this question.

I would like to hear others opinion on this though.

Reply:
Maybe this question shouldn't even be in the survey (sometimes I use .catch, try catch and "leave to others" all in the same code).

Reply:
Kept the question and expanded the "leave handling..." answer.